packages/cli/src/services/test-webhook-registrations.service.ts

```diff
@@ -0,0 +1,102 @@
+import { Service } from 'typedi';
+import { CacheService } from './cache.service';
+import { ApplicationError, type IWebhookData } from 'n8n-workflow';
+import type { IWorkflowDb } from '@/Interfaces';
+
+export type TestWebhookRegistration = {
+	sessionId?: string;
+	workflowEntity: IWorkflowDb;
+	destinationNode?: string;
+	webhook: IWebhookData;
+};
+
+@Service()
+export class TestWebhookRegistrationsService {
+	constructor(private readonly cacheService: CacheService) {}
+
+	private readonly cacheKey = 'test-webhook';
+
+	async register(registration: TestWebhookRegistration) {
+		const key = this.toKey(registration.webhook);
+
+		await this.cacheService.set(key, registration);
+	}
+
+	async deregister(arg: IWebhookData | string) {
+		if (typeof arg === 'string') {
+			await this.cacheService.delete(arg);
+		} else {
+			const key = this.toKey(arg);
+			await this.cacheService.delete(key);
+		}
+	}
+
+	async exists(key: string) {
+		const registration = await this.cacheService.get<TestWebhookRegistration>(key);
+
+		return registration !== undefined;
+	}
+
+	async get(key: string) {
+		const registration = await this.cacheService.get<TestWebhookRegistration>(key);
+
+		if (!registration) {
+			throw new ApplicationError('Failed to find test webhook registration', { extra: { key } });
+		}
+
+		return registration;
+	}
+
+	async getAllKeys() {
+		const keys = await this.cacheService.keys();
+
+		const CACHE_PREFIX = 'n8n:cache';
+		const FULL_CACHE_KEY = `${CACHE_PREFIX}:${this.cacheKey}`;
+
+		return keys
+			.filter((key) => key.startsWith(FULL_CACHE_KEY))
+			.map((key) => key.slice(`${CACHE_PREFIX}:`.length));
+	}
+
+	async getAllRegistrations() {
+		const keys = await this.getAllKeys();
+
+		return this.cacheService.getMany<TestWebhookRegistration>(keys);
+	}
```

Comment:
The two operations here are somewhat costly, especially if we use `cachesService.keys()` that returns all keys. According to [Redis Docs](https://redis.io/commands/keys/)

> Warning: consider KEYS as a command that should only be used in production environments with extreme care. It may ruin performance when it is executed against large databases. This command is intended for debugging and special operations, such as changing your keyspace layout. Don't use KEYS in your regular application code. If you're looking for a way to find keys in a subset of your keyspace, consider using [SCAN](https://redis.io/commands/scan) or [sets](https://redis.io/topics/data-types#sets).


Have we considered using [Redis Hashes](https://redis.io/docs/data-types/hashes/) instead? So it's always a single key.

We could also atomically add and remove data in place, without downsides.

Reply:
Redis Hashes are pending and will be added soon here: https://linear.app/n8n/issue/PAY-1188/improve-cache-service